src/WebJobs.Extensions.Http/AuthenticatedUserBindingProvider.cs

```diff
@@ -0,0 +1,168 @@
+                    _bindingData = bindingData;
+                }
+
+                public Type Type => typeof(AuthenticatedUser);
+
+                public async Task<object> GetValueAsync()
+                {
+                    var initialRequest = _bindingData.Values.FirstOrDefault(val => val.GetType() == typeof(HttpRequestMessage)) as HttpRequestMessage;
+                    if (initialRequest == null)
+                    {
+                        throw new InvalidOperationException("Cannot determine authenticated user without HttpTrigger.");
+                    }
+
+                    var authorizationHeader = GetAuthorizationHeaderValue(initialRequest);
+                    if (authorizationHeader != null)
+                    {
+                        return await GetAuthenticatedUserFromEasyAuth(authorizationHeader);
+                    }
+                    else
+                    {
+                        return GetAuthenticatedUserFromFunctionKey();
+                    }   
+                }
+
+                private static string GetAuthorizationHeaderValue(HttpRequestMessage initialRequest)
+                {
+                    var idTokenHeaders = initialRequest.Headers.Where(header => header.Key.EndsWith("id-token", StringComparison.OrdinalIgnoreCase));
+                    if (!idTokenHeaders.Any())
+                    {
+                        return null;
+                    }
+                    else
+                    {
+                        var idTokenHeader = idTokenHeaders.First();
+                        var idTokenValue = idTokenHeader.Value.First();
+                        return "Bearer " + idTokenValue;
+                    }
+                }
+
+                private async Task<AuthenticatedUser> GetAuthenticatedUserFromEasyAuth(string authorizationHeader)
+                {
+                    string hostname = Environment.GetEnvironmentVariable("WEBSITE_HOSTNAME");
+                    var authUri = "https://" + hostname + "/.auth/me";
+                    HttpRequestMessage request = new HttpRequestMessage(HttpMethod.Get, authUri);
+                    request.Headers.Add("Authorization", authorizationHeader);
+                    HttpResponseMessage response = await _client.SendAsync(request);
+
+                    if (response.StatusCode == HttpStatusCode.Unauthorized)
+                    {
+                        // TODO: Determine if an exception is the most appropriate use case here
+                        // especially once we start dealing with AND/OR of multiple AuthLevel enums
+                        throw new InvalidOperationException("No user was authenticated.");
+                    }
+                    else if (response.StatusCode == HttpStatusCode.NotFound)
+                    {
+                        //Same concern as the unauthorized case
+                        throw new InvalidOperationException("Authentication/Authorization is not enabled.");
```

Comment:
Good point about the token store. For now, we have all of the required fields except for the claims (and we want to add those in a header soon anyways). Once we have all of the fields present, then this can become the only method used to populate, and we avoid hitting the token store altogether.

---
In reply to: [171931203](https://github.com/Azure/azure-webjobs-sdk-extensions/pull/373#discussion_r171931203) [](ancestors = 171931203)